### DISCUSSION TITLE
Creation of sa.Table with large number of columns

### DISCUSSION THREAD

**Discussion by saumzzz** [sqlalchemy/sqlalchemy#10111] [Usage Questions]
My usecase requires creating non-dbms related tables, which won't be used for ddl/dml. These tables however have relationships with each other.
I'm leveraging sa.Table to create in-memory tables, and one of the tables I'm creating has ~5000 columns.
The `sa.Table(alias, meta, *cols)` is taking a lot of time.

Any ideas on how to speed this up?

**Comment by zzzeek** [comment]
I would basically not create a `Table` object with 5000 `Column` objects in the first place.  If you need to write a SELECT thats against, say, 100 of those columns, then just make the `Table` with only those `Column` objects that you need for queries.  if you aren't emitting DDL then you can ignore the other 4900 columns.

**Reply by saumzzz** [reply]
Agreed, I'm using the `sa.Table` and `Metadata` to model actual data tables and the relationships between them, these tables do have 5000+ columns with data and hence their sqlalchemy counterparts also need to have the columns. 
The perf times do become faster with sqlalchemy v2.x but unfortunately I'm stuck with v1.4.x

any suggestions to optimize this would be really helpful

**Reply by CaselIT** [reply]
if you don't need the full `sa.Table` features you may be able to use the lovercase version of table and columns that may be faster: `sa.table` and `sa.column`

**Reply by CaselIT** [reply]
also for reference creating a table with 5k columns takes about 40ms on my pc, that does not seem slow.
of what timing are we talking about here?

**Reply by saumzzz** [reply]
With 1.4.7 the time taken on my pc is ~20s
With 2.0.2 the time taken on my pc is ~400ms

**Reply by CaselIT** [reply]
there times seem very strange if creating the table in python.
Are you reflecting the table?
That's not mentioned anywhere, so I just assumed you were just creating the columns in python

**Reply by saumzzz** [reply]
Yes that's right, I'm not reflecting the table I'm just creating the columns in both cases (v1.4.47, v2.0.2)

**Reply by saumzzz** [reply]
Also, made a mistake here https://github.com/sqlalchemy/sqlalchemy/discussions/10111#discussioncomment-6476247 with the version 
it's supposed to be v1.4.47
Apologies, my bad!

**Comment by saumzzz** [comment]
A workaround I used to optimize is to use, do you think this would cause any issues?

```python
constraint_cols = [...] # index, unique, pkey, fkey, etc. 
non-constraint_cols = [...]


table = sa.Table(name, meta, *constraint_cols)
table._columns.extend(<non-constraint-columns>) # instead of the `replace` method
# further handling of non-constraint cols
```

**Reply by CaselIT** [reply]
btw prefer append column https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table.append_column

what you are doing may not completely register the column

**Reply by CaselIT** [reply]
Oh sorry, I did bot read the first line.
If the column does not have any default/constraint/index/etc (meaning it just defines name and type) that may work

**Comment by AbdealiLoKo** [comment]
@CaselIT There is definitely been some changes in performance that I can reproduce.

sqlalchemy 1.4.x was somewhat slow  -- 7sec
Later in sqlalchemy 2.0.x - it became much slower 🐌  -- 14sec
But latest sqlalchemy 2.0.19 - is blazing fast 🚀  -- 0.5sec

Just some observations that may help others.

MRE:
```
$ docker run --rm -it ubuntu:22.04 bash
# apt update
# apt install -y python3 python3-pip
# python3 -V
Python 3.10.6
# pip3 install sqlalchemy==1.4.7
# time python3 -c "import sqlalchemy as sa; sa.Table('abc', sa.MetaData(), *[sa.Column(f'c_{i}') for i in range(5000)])"
real    0m6.746s
user    0m5.360s
sys     0m0.352s

# pip3 install sqlalchemy==1.4.47
# time python3 -c "import sqlalchemy as sa; sa.Table('abc', sa.MetaData(), *[sa.Column(f'c_{i}') for i in range(5000)])"
real    0m7.773s
user    0m5.712s
sys     0m0.483s

# pip3 install sqlalchemy==2.0.2
# time python3 -c "import sqlalchemy as sa; sa.Table('abc', sa.MetaData(), *[sa.Column(f'c_{i}') for i in range(5000)])"
real    0m14.446s
user    0m12.296s
sys     0m0.499s

# pip3 install sqlalchemy==2.0.19
# time python3 -c "import sqlalchemy as sa; sa.Table('abc', sa.MetaData(), *[sa.Column(f'c_{i}') for i in range(5000)])"
real    0m0.519s
user    0m0.402s
sys     0m0.044s
```

**Reply by CaselIT** [reply]
Ok this seems like this change https://docs.sqlalchemy.org/en/20/changelog/changelog_20.html#change-04c73bfefd3c83673f8d42f5f0eb2d8c on 2.0.14

I did not think it was that effective :)

Regarding options
- try if lowercase table is an option (I have not checked if it has the same performance as Table)
- update to v2
- specify only the columns you actually need
